tests/drivers/i2c/i2c_emul/src/test_forwarding_pio.cpp

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright 2024 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "emulated_target.hpp"
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/fff.h>
+#include <zephyr/ztest.h>
+
+namespace
+{
+
+#define GET_TARGET_DEVICE(node_id, prop, n) DEVICE_DT_GET(DT_PHANDLE_BY_IDX(node_id, prop, n)),
+
+/* Get the devicetree constants */
+constexpr const struct device *controller = DEVICE_DT_GET(CONTROLLER_LABEL);
+constexpr const struct device *targets[FORWARD_COUNT] = {
+	DT_FOREACH_PROP_ELEM(CONTROLLER_LABEL, forwards, GET_TARGET_DEVICE)};
+
+ZTEST(i2c_emul_forwarding, test_write_is_forwarded)
+{
+	// Try writing some values
+	for (uint8_t data = 0; data < 10; ++data) {
+		const int expected_call_count = 1 + data;
+
+		zassert_ok(i2c_write(controller, &data, sizeof(data),
+				     emulated_target_config[0].address));
+
+		// Expected no reads to be made
+		zexpect_equal(0, target_read_requested_0_fake.call_count);
+		zexpect_equal(0, target_read_processed_0_fake.call_count);
+
+		// Expected N write requests to be made
+		zexpect_equal(expected_call_count, target_write_requested_0_fake.call_count,
+			      "Expected to be called %d times, got %d", expected_call_count,
+			      target_write_requested_0_fake.call_count);
+		zexpect_equal(expected_call_count, target_write_received_0_fake.call_count,
+			      "Expected to be called %d times, got %d", expected_call_count,
+			      target_write_received_0_fake.call_count);
+
+		// Check that the data written was correct
+		zexpect_equal(data, target_write_received_0_fake.arg1_val,
+			      "Expected data value %u. got %u", data,
+			      target_write_received_0_fake.arg1_val);
+
+		// Expect 1 stop call per write request
+		zexpect_equal(expected_call_count, target_stop_0_fake.call_count,
+			      "Expected to be called %d times, got %d", expected_call_count,
+			      target_stop_0_fake.call_count);
+	}
+}
+
+ZTEST(i2c_emul_forwarding, test_read_is_forwarded)
+{
+	// Try reading some values
+	for (uint8_t i = 0; i < 10; ++i) {
+		const uint8_t expected_data[2] = {
+			static_cast<uint8_t>(0x1 * i),
```

Comment:
can't we just cast? (uint8_t)(0x1*i), easy peasy

Reply:
I guess, but why not use the static_cast which verifies that the value can be cast statically? The main up side of the static_cast is that if we go back and change `i` or the constant we're multiplying by to ones that wouldn't be compatible with casting to a uint8, we would catch it at compile time. Overall the reason is just future proofing the cast.

Reply:
I don't really know what this means by future proofing here, a cast is a cast? The result type is uint8_t and well known here. Surprising to even see the need for a cast to be honest. Not a deal breaker for me, but was confusing to see.